server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstraction.java

```diff
@@ -67,6 +67,12 @@
     @Nullable
     IndexMetadata getWriteIndex();
 
+    /**
+     * @return the data stream to which this index belongs or <code>null</code> if this is not a concrete index or
+     * if it is a concrete index that does not belong to a data stream.
+     */
+    @Nullable DataStream getParentDataStream();
```

Comment:
I chose the return type of `IndexAbstraction.DataStream` for two reasons -- the parent data stream could never be an `IndexAbstraction.Index` or `IndexAbstraction.Alias` and when implementing data stream rollover in the next PR, I added a `getDataStream` method to `IndexAbstraction.DataStream` which is convenient to call from `getParentDataStream` without casting.

Reply:
Ok, that makes sense. I just prefer that other code isn't going to rely on data stream specific methods (like is today with the alias impl). Otherwise the abstraction is going to be leaky.